README.md
fastavro - Now even faster!
===========================

The Python [`avro`][avro_pypi] package distributed by 
[Apache][apache_avro] is **incredibly slow**. When dealing with very 
large datasets, the performance of [`avro`][avro_pypi] is unacceptable,
more or less unfit for any production use.

**`fastavro`** is less feature complete than [`avro`][avro_pypi], 
however performance reading and writing Avro files is significantly
faster.

With the [Cython][cython] extension modules compiled, **`fastavro`**
is **really** fast, almost on par with the native Java implementation
of Avro.

The focus of this fork ([e-heller/fastavro]) is on reimplementing and 
optimizing `fastavro's` [Cython][cython] extension modules in *pure
Cython code* to achieve screaming speed.

I hope these improvements will eventually be pulled into the upstream 
repo, but that may take some time.

This **`fastavro`** fork supports these Python versions:

* Python 2.6
* Python 2.7
* Python 3.4
* Python 3.5


[e-heller/fastavro]: https://github.com/e-heller/fastavro
[avro_pypi]: https://pypi.python.org/pypi/avro
[apache_avro]: http://avro.apache.org
[Cython]: http://cython.org


Build and Install
=================

Because **`fastavro`** is shipped with [Cython][cython] extension
modules, you will require the following to build and install:

1. [Cython][cython_pypi] to generate the C extension files

2. Some kind of C compiler like `gcc`

If you can't compile the extension modules, you can still use the
*pure Python* implementation, but you will be missing out on the
significant performance improvements.

Compile the source
------------------

First, download [Cython][cython_pypi]:

```shell
$ pip install cython
```

To build the compiled extensions, from the root `fastavro` source
directory:

```shell
$ make cfiles
$ python setup.py build
```

Install the package
-------------------

Assuming the build worked, now you can just type:

```shell
$ python setup.py install
```


Basic Usage
===========

Reading from Avro format
------------------------

```python
import fastavro

with open('some_file.avro', 'rb') as input:
    # Create a `Reader` object
    reader = fastavro.Reader(input)

    # Obtain the writer's schema if required
    schema = reader.schema

    # Iteratively read the records:
    for record in reader:
        process_record(record)
        
    # Read the records in one shot:
    records = list(reader)
```


Writing to Avro format
----------------------

```python
import fastavro

# Define an Avro Schema as a dict:
schema = {
    'doc': 'A weather reading.',
    'name': 'Weather',
    'namespace': 'test',
    'type': 'record',
    'fields': [
        {'name': 'station', 'type': 'string'},
        {'name': 'time', 'type': 'long'},
        {'name': 'temp', 'type': 'int'},
    ],
}

# Create some records:
records = [
    {u'station': u'011990-99999', u'temp': 0, u'time': 1433269388},
    {u'station': u'011990-99999', u'temp': 22, u'time': 1433270389},
    {u'station': u'011990-99999', u'temp': -11, u'time': 1433273379},
    {u'station': u'012650-99999', u'temp': 111, u'time': 1433275478},
]

# Open a file, and write the records:
with open('some_file.avro', 'wb') as out:
    fastavro.write(out, schema, records)
```


Limitations
===========

**`fastavro`** is missing many of the features of the official Apache
[`avro`][avro_pypi] package. Essentially, **`fastavro`** only supports
reading and writing in Avro format.

Notably, there is no support for:

* [Avro Protocols][spec_protocol]

* The [Protocol Wire Format][spec_wire_format] or any kind of Wire
  Transmission

There are also some limitations with reading and writing:

* Incomplete support for *'reader's schemas'* - i.e., reading an Avro
  file written with one schema (the 'writer's schema') and interpreting
  the data with another schema (the 'reader's schema').
  See [Schema Resolution][spec_schema_res] in the Avro specification
  for details.
  
* No support for *Aliases* - i.e., the `aliases` attribute in named 
  Avro types. See [Aliases][spec_aliases] in the Avro specification
  for details.
  
* No support for *Logical Types* - i.e., the new `logicalType` attribute
  that adds support for derived types in Avro 1.8 like `decimal`,
  `date`, `time-millis`, `timestamp-millis`, etc.
  See [Logical Types][spec_logical] in the Avro specification for
  details.
  > Note: the [ViaSat/fastavro] fork currently has support for some
  Logical Types. I hope to incorporate this work soon.


[spec_wire_format]: https://avro.apache.org/docs/current/spec.html#Protocol+Wire+Format
[spec_protocol]: https://avro.apache.org/docs/current/spec.html#Protocol+Declaration
[spec_schema_res]: https://avro.apache.org/docs/current/spec.html#Schema+Resolution
[spec_aliases]: https://avro.apache.org/docs/current/spec.html#Aliases
[spec_logical]: https://avro.apache.org/docs/current/spec.html#Logical+Types
[ViaSat/fastavro]: https://github.com/ViaSat/fastavro/


Hacking
=======

If you want to play around and modify any of the [Cython][cython]
`.pyx` files, you will need to recompile the C module code.

You will require:

1. [Cython][cython_pypi] to regenerate the C files
2. Some kind of C compiler

To easiest way to recompile the C extensions is as follows.

From the root `fastavro` source directory:

```shell
$ make build
```

The `make build` command first calls `cython` to generate the C files,
then compiles them via `setup.py`. (The exact command is
`python setup.py build_ext -i`)

If you are feeling more adventurous, and you're on a Linux-y platform
with `gcc` you can try to compile the extensions with the GCC settings
I personally use:

```shell
$ make compile
```

This will compile the extensions "in place" in the source directory.

To install the package with your modifications, you'll just have to 
manually copy the `fastavro` package directory to your Python
`site-packages` directory.

I definitely recommend working inside a virtual environment if you plan 
to modify and rebuild `fastavro`. If you are unfamiliar with virtual
environments, check out [virtualenv][venv] or [pew][pew].
Personally, I recommend [pew][pew].


[cython]: http://cython.org/
[cython_pypi]: https://pypi.python.org/pypi/Cython
[venv]: http://pypi.python.org/pypi/virtualenv
[pew]: https://pypi.python.org/pypi/pew
[distutils]: https://docs.python.org/2.7/extending/building.html


Changes
=======

See the [ChangeLog]

[ChangeLog]: https://github.com/e-heller/fastavro/blob/master/ChangeLog


Contact
=======

[The e-heller/fastavro fork](https://github.com/e-heller/fastavro/)

[Official fastavro repo](https://github.com/tebeka/fastavro)


ChangeLog
2016-05-15 version 0.9.10
    * Reorganized most files, new docstrings and comments
    * the `six` module has been renamed to `compat` to avoid confusion with
      the real 'six' package.
    * Most schema-related definitions and functions have been moved to the
      `schema` module for logical consistency.
    * `iter_avro` class has been renamed to `Reader`, and the `_iter_avro`
      function is removed with its implementation now in the `Reader` class.
    * Reimplemented `match_schemas` function is properly recursive, and returns
      True of False -- does not raise any exceptions.
    * Added promotion cases for 'bytes' -> 'string' and 'string' -> 'bytes'
      per Avro specification docs.
    * `acquaint_schema` now rejects schemas with missing required attributes.
      E.g., a 'record' requires a 'name' attribute. A 'record' without a 'name'
      is now rejected and a `SchemaError` is raised.
    * Small optimizations here and there (a little benchmarking showed around
      15-20% improvement running pure python code)
    * A few new test cases
    * All changes @e-heller

2016-02-13 version 0.9.9
    * Raise EOF on empty input
    * numpy style docs

2016-01-15 version 0.9.8
    * Add some docstrings (issue #45)
    * Don't use mutable default arguments

2015-12-27 version 0.9.7
    * Python 2.6 support (@antonymayi in PR #44)
    * Some code cleanup

2015-10-14 version 0.9.6
    * Support streaming IO (@mjwillson in PR #40)

2015-10-04 version 0.9.5
    * Fix issue with writing block (issue #35)

2015-09-01 version 0.9.4
    * Support schema migration (@scottbelden in PR #31)

2015-09-01 version 0.9.3
    * Speedup (@scottbelden in PR #30)
    * Fix writer.validate on Python 3 (@mjwillson in PR #29)

2015-08-25 version 0.9.2
    * allow extra metadata to be used (@scottbelden in PR #28)

2015-08-21 version 0.9.1
    * Fix encoding of fields with default values (@scottbelden in PR #25)
    * Fix boolean decoding (@scottbelden in PR #26)

2015-08-20 version 0.9.0
    * Handle error types (@scottbelden in PR #20)
    * Fix boolean encoding/decoding (@rodcarroll in PR #22)
    * Support binary encoder (issue #14, @scottbelden in PR #24)

2015-08-18 version 0.8.8
    * handle records with no fields (@scottbelden in PR #19)

2015-08-15 version 0.8.7
    * fix boolean encoding (@scottbelden in PR #18)

2015-08-13 version 0.8.6
    * Allow nested record definitions from arrays and maps (@mtth in PR #16)
    * Fix namespace bug for union references. (@mtth in PR #16)
    * Remove generated .c files from git
    * Testing with Python 3.5rc1

2015-08-03 version 0.8.5
    * Fixed nested schema bug (@dcreager in PR #15)
    * Testing Python 3.5 and pypy3
    * Don't build extension on *any* version of pypy

2015-08-03 version 0.8.4
    * Packing fixes
    * Detect unknown types (@pkoch in PR #11)
    * Fix .decode with no codec (@oliverbestmann in PR #9)

2015-07-14 version 0.8.3
    * Fixed pip install manifest issue (@rouge8 in PR #8)

2015-07-14 version 0.8.2
    * Main repo moved to github

2015-06-02 version 0.8.1
    * Fix float/double (@rodcarool in github #5)
    * Write documentation

2015-05-04 version 0.8.0
    * Write support (Paulo Köch in bitbucket PR #4)

2015-04-28 version 0.7.10
    * Unexpose `write` (ref from bitbucket #2)

2014-08-28 version 0.7.9
    * Expose "load" (github #1)

2013-04-20 version 0.7.8
    * Don't compile extension under pypy > 1.8 (issue #12)
    * pyflakes cleanup

2013-03-27 version 0.7.7
    * Fixed bug when stdout.encoding is None

2013-03-26 version 0.7.6
    * Fixed regression with --schema

2013-03-23 version 0.7.5
    * Fixed regression with Cython and multiple definitions

2013-03-01 version 0.7.4
    * Fixed regression in 2.7 and recursive

2013-02-19 version 0.7.3
    * Fixed bug in Python 3 printing

2012-12-10 version 0.7.2
    * Fixed bug in argument parsing

2012-12-10 version 0.7.1
    * Fixed README (about snappy)

2012-12-10 version 0.7.0
    * Snappy support

2012-10-05 version 0.6.10
    * Handle names in unions (aboSamoor)
    * Handle namespaces (aboSamoor)

2012-10-05 version 0.6.9
    * Fixed handling if named records (issue #5)

2012-07-13 version 0.6.8
    * Added "next" method to reader

2012-04-29 version 0.6.7
    * Support recursive schema (issue #4)

2012-04-28 version 0.6.6
    * Using Cython on six as well
    * fastavro/pyfastavro.py -> fastavro/reader.py
    * Show more info when running tests

2012-04-24 version 0.6.5
    * Better error message when file is not avro file
    * Fixed bug in error message when can't read file

2012-03-16 version 0.6.4
    * Python 2.6 compatible (issue #3, thanks trochee)

2012-03-16 version 0.6.3
    * Faster (using cython: auto_cpdef=True), thanks to Stefan Behnel

2012-03-10 version 0.6.2
    * Simpler iterator
    * Trove categories

2012-03-10 version 0.6.1
    * Python 3 support works
    * Better testing

2012-03-10 version 0.6.0
    * Python 3 support

2012-02-22 version 0.5.0
    * Generate `fastavro` script

2012-01-30 version 0.4.2
    * Compliance with Apache license

2012-01-25 version 0.4.1
    * Several files in command line
    * Output encoding
    * Conditional build of extension (from simplejson)

2012-01-25 version 0.4.0
    * fastavro.reader
    * Include README.rst in sdist package

2012-01-24 version 0.3.2
    * Fixed read_array (issue #2)

2012-01-24 version 0.3.1
    * Fixed read_enum (issue #1)

2012-01-23 version 0.3.0
    * Distribute generate C file (idea by Juan)

2012-01-11 version 0.2.2
    * Long description
    * Get version from fastavro/__init__.py

2012-01-11 version 0.2.1
    * --schema

2012-01-11 version 0.2.0
    * Optionally use Cython (more speedup)

2012-01-05 version 0.1.0
    * Initial release


run-tests.sh
#!/bin/bash
# Run tests suite, you can specify which nose to run by setting environment
# variable `nose` (running without will use `nosetests`).
#   nose=nosetests-3.2 ./run-tests.sh

# Exit on error
# set -e

echo "[$(date +%Y%m%dT%H%M%S)] ${USER}@$(hostname) :: $(python --version 2>&1)"
echo

echo "running flake8"
flake8 fastavro tests
if [ $? -eq 0 ]
then echo "OK"
else echo "FAILED"
fi
echo

nose=${nose-nosetests}
echo "running $nose"
${nose} -vd $@ tests


tests/utils.py
# -*- coding: utf-8 -*-
"""Misc utlities for fastavro tests"""

from __future__ import absolute_import

# flake8: noqa

import random
import struct
import sys


PY2 = sys.version_info[0] == 2
PY3 = sys.version_info[0] == 3

if PY2:
    _unicode_type = unicode
    _bytes_type = str
    unichr = unichr
    xrange = xrange
else:
    _unicode_type = str
    _bytes_type = bytes
    unichr = chr
    xrange = range


_byte_pack = struct.Struct('>B').pack


def int2byte(n):
    """Equivalent to chr(n) in Python 2. Works for Python 3"""
    return chr(n) if PY2 else _byte_pack(n)


def iteritems(obj):
    """This is compatible for Python 2 or 3"""
    return obj.iteritems() if PY2 else iter(obj.items())


# ---- Generating random strings ---------------------------------------------#

def random_byte_str(count):
    """Create a random byte string (may be non-ascii)"""
    return bytes(b''.join(
        int2byte(random.randint(1, 255)) for _ in xrange(count)
    ))


def random_unicode_str(count):
    """Create a random utf-8 compatible unicode string"""
    # Max UTF-8 code point is 0xFFFF or whatever this Python supports
    max_code_pt = min(0xFFFF, sys.maxunicode)

    def _code_pt():
        p = random.randint(1, max_code_pt)
        # Ordinals in this range are surrogates, invalid for utf-8 encoding
        while 0xD000 <= p <= 0xDFFF:
            p = random.randint(1, max_code_pt)
        return p

    return u''.join(
        unichr(_code_pt()) for _ in xrange(count)
    )


tests/__init__.py
# -*- coding: utf-8 -*-
"""Tests for fastavro"""

from __future__ import absolute_import

from pkgutil import extend_path
__path__ = extend_path(__path__, 'fastavro')

import fastavro  # noqa


fastavro/compat.py
# -*- coding: utf-8 -*-
"""Compatiblity for Python versions"""


# Some of this code is adapted from `CherryPy`:
#   https://bitbucket.org/cherrypy/cherrypy/wiki/Home
#   Licensed under a BSD License:
#   https://bitbucket.org/cherrypy/cherrypy/src/tip/cherrypy/LICENSE.txt

# Some of this code is adapted from `six`:
#   https://bitbucket.org/gutworth/six
#   Licensed under a MIT License:
#   https://bitbucket.org/gutworth/six/src/tip/LICENSE


from __future__ import absolute_import

# flake8: noqa

import sys


PY2 = sys.version_info[0] == 2
PY3 = sys.version_info[0] == 3

if PY2:
    # Python 2 type aliases
    _unicode_type = unicode
    _bytes_type = str
    _string_types = (basestring,)
    _int_types = (int, long)
    _number_types = (int, long, float)

    from cStringIO import StringIO as BytesIO
    xrange = xrange

    def py2_iteritems(obj):
        return obj.iteritems()
else:
    # Python 3 type aliases
    _unicode_type = str
    _bytes_type = bytes
    _string_types = (bytes, str)
    _int_types = (int,)
    _number_types = (int, float)

    from io import BytesIO
    xrange = range

    def py3_iteritems(obj):
        return iter(obj.items())


# Export an alias for each of the version-specific functions
# We do it this way because Cython does not like redefined functions.
if PY2:
    iteritems = py2_iteritems
else:
    iteritems = py3_iteritems


def byte2int(b):
    return ord(b[0]) if PY2 else b[0]


fastavro/__main__.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Command-line tool to read an Avro file and emit the records in JSON format
to stdout.
"""

from __future__ import absolute_import

import sys

try:
    import simplejson as json
except ImportError:
    import json

import fastavro


def _json_dump(obj, indent, output=None):
    output = output or sys.stdout
    # Try to get encoding from output (eg `sys.stdout.encoding`)
    # Default to 'utf-8' if None
    encoding = getattr(output, 'encoding', None) or 'utf-8'
    json.dump(obj, output, indent=indent, encoding=encoding)


def main(argv=None):
    from argparse import ArgumentParser

    argv = argv or sys.argv

    parser = ArgumentParser(
        description='Reads an Avro file, emitting the records as JSON')
    parser.add_argument('file', help='file(s) to parse', nargs='*')
    parser.add_argument('--schema', help='dump schema instead of records',
                        action='store_true', default=False)
    parser.add_argument('--codecs', help='print supported codecs',
                        action='store_true', default=False)
    parser.add_argument('--version', action='version',
                        version='fastavro %s' % fastavro.__version__)
    parser.add_argument('-p', '--pretty', help='pretty print json',
                        action='store_true', default=False)
    args = parser.parse_args(argv[1:])

    if args.codecs:
        print('\n'.join(sorted(fastavro._reader.BLOCK_READERS)))
        raise SystemExit

    files = args.file or ['-']
    for filename in files:
        if filename == '-':
            fo = sys.stdin
        else:
            try:
                fo = open(filename, 'rb')
            except IOError as e:
                raise SystemExit('error: cannot open %s - %s' % (filename, e))

        try:
            reader = fastavro.reader(fo)
        except ValueError as e:
            raise SystemExit('error: %s' % e)

        if args.schema:
            _json_dump(reader.schema, True)
            sys.stdout.write('\n')
            continue

        indent = 4 if args.pretty else None
        try:
            for record in reader:
                _json_dump(record, indent)
                sys.stdout.write('\n')
        except (IOError, KeyboardInterrupt):
            pass


if __name__ == '__main__':
    main()


fastavro/writer.py
# -*- coding: utf-8 -*-
"""Python implementation for writing Apache Avro files"""


# This code is based on the Apache 'avro' pacakge, found at:
# http://svn.apache.org/viewvc/avro/trunk/lang/py/src/avro/
# Under the Apache 2.0 License (https://www.apache.org/licenses/LICENSE-2.0)

# Please refer to the Avro specification page for details on data encoding:
#   https://avro.apache.org/docs/current/spec.html


from __future__ import absolute_import

from binascii import crc32
from collections import Iterable, Mapping
from os import urandom
from struct import pack
from zlib import compress

try:
    import simplejson as json
except ImportError:
    import json

try:
    import snappy
except ImportError:
    snappy = None

from fastavro.compat import (
    BytesIO, iteritems, _int_types, _number_types, _unicode_type,
    _bytes_type, _string_types,
)
from fastavro.schema import (
    extract_named_schemas_into_repo, HEADER_SCHEMA, MAGIC, SYNC_SIZE,
    SYNC_INTERVAL, PRIMITIVE_TYPES,
)


INT_MIN_VALUE = -(1 << 31)
INT_MAX_VALUE = (1 << 31) - 1
LONG_MIN_VALUE = -(1 << 63)
LONG_MAX_VALUE = (1 << 63) - 1


class _NoValue(object):
    def __repr__(self):
        return '<NoValue>'

    def __str__(self):
        return repr(self)

NoValue = _NoValue()


# ---- Writing Avro primitives -----------------------------------------------#

def write_null(stream, datum, schema=None):
    """A `null` value is not written at all."""
    pass


def write_boolean(stream, datum, schema=None):
    """A `boolean` value is written as a single byte: b'0x00' for False,
    b'0x01' for True.

    Reference: https://avro.apache.org/docs/current/spec.html#binary_encode_primitive
    """  # noqa
    stream.write(b'\x01' if datum else b'\x00')


def write_long(stream, datum, schema=None):
    """An `int` or `long` value is written as 32-bit or 64-bit integer,
    respectively, using a variable-length zig-zag encoding. This is the same
    encoding used in Google protobufs.

    A good explanation of this encoding can be found at:
        https://developers.google.com/protocol-buffers/docs/encoding#signed-integers

    Reference: https://avro.apache.org/docs/current/spec.html#binary_encode_primitive
    """  # noqa
    datum = (datum << 1) ^ (datum >> 63)
    while (datum & ~0x7F) != 0:
        stream.write(pack('B', (datum & 0x7f) | 0x80))
        datum >>= 7
    stream.write(pack('B', datum))


# Alias `write_int` to `write_long`
write_int = write_long


def write_float(stream, datum, schema=None):
    """A `float` value is written as a single precision 32-bit IEEE 754
    floating-point value in little-endian format.

    See the implementation of `_PyFloat_Pack4()` in Python's floatobject.c

    Reference: https://avro.apache.org/docs/current/spec.html#binary_encode_primitive
    """  # noqa
    stream.write(pack('<f', datum))


def write_double(stream, datum, schema=None):
    """A `double` value is written as a double precision 64-bit IEEE 754
    floating-point value in little-endian format.

    See the implementation of `_PyFloat_Pack8()` in Python's floatobject.c

    Reference: https://avro.apache.org/docs/current/spec.html#binary_encode_primitive
    """  # noqa
    stream.write(pack('<d', datum))


def write_bytes(stream, datum, schema=None):
    """A `bytes` value is written as a `long` (length of the byte string),
    immediately followed by the raw byte data.

    Reference: https://avro.apache.org/docs/current/spec.html#binary_encode_primitive
    """  # noqa
    write_long(stream, len(datum))
    stream.write(datum)


def write_string(stream, datum, schema=None):
    """A `string` value is written as a `long` (length of the UTF-8 encoded
    string), immediately followed by the UTF-8 encoded byte data.

    Note: Avro `string` values *must* be encoded to UTF-8 byte strings.

    Reference: https://avro.apache.org/docs/current/spec.html#binary_encode_primitive
    """  # noqa
    byte_str = (
        datum.encode('utf-8') if isinstance(datum, _unicode_type) else datum
    )
    write_long(stream, len(byte_str))
    stream.write(byte_str)


# ---- Writing Avro complex types --------------------------------------------#

def write_fixed(stream, datum, schema=None):
    """A `fixed` value is written as raw bytes. The length of the byte data
    is declared in the schema.

    Note: we do not verify that the length of `datum` matches the length
    declared in the `schema` (although perhaps we should). For now, this
    responsibility is left to the user.

    Reference: https://avro.apache.org/docs/current/spec.html#Fixed
    """
    stream.write(datum)


def write_enum(stream, datum, schema):
    """An `enum` value is written as an `int` representing the zero-based
    position of the symbol in the schema.

    Note: calling `.index(datum)` repeatedly is a bit inefficient. Instead, we
    could precompute the index-offsets for each symbol and store it in a dict.

    Reference: https://avro.apache.org/docs/current/spec.html#Enums
    """
    index = schema['symbols'].index(datum)
    write_int(stream, index)


def write_array(stream, datum, schema):
    """An `array` value is written as a series of blocks.

    Each block consists of a `long` `count` value, followed by that many array
    items. A block with `count` zero indicates the end of the array. Each item
    is encoded per the array's item schema.

    If a block's `count` is negative, then the `count` is followed immediately
    by a `long` block size, indicating the number of bytes in the block. The
    actual `count` in this case is the absolute value of the `count` written.

    Note: this implementation does not support writing negative block counts.

    Reference: https://avro.apache.org/docs/current/spec.html#Arrays
    """
    datum_len = len(datum)
    if datum_len:
        write_long(stream, datum_len)
        dtype = schema['items']
        for item in datum:
            write_data(stream, item, dtype)
    write_long(stream, 0)


def write_map(stream, datum, schema):
    """A `map` value is written as a series of blocks.

    Each block consists of a `long` `count` value, followed by that many
    key / value pairs. A block with `count` zero indicates the end of the map.

    If a block's `count` is negative, then the `count` is followed immediately
    by a `long` block size, indicating the number of bytes in the block. The
    actual `count` in this case is the absolute value of the `count` written.

    For writing the key / value pairs:
        A map `key` is assumed to be string
        A map `value` is written per the map's 'value' schema

    Note: this implementation does not support writing negative block counts.

    Reference: https://avro.apache.org/docs/current/spec.html#Maps
    """
    datum_len = len(datum)
    if datum_len:
        write_long(stream, datum_len)
        vtype = schema['values']
        for key, val in iteritems(datum):
            write_string(stream, key)
            write_data(stream, val, vtype)
    write_long(stream, 0)


def write_union(stream, datum, schema):
    """A `union` value is written as a `long` indicating the zero-based
    position of the `value` in the union's schema, immediately followed by
    the `value`, written per the indicated schema within the union.

    Reference: https://avro.apache.org/docs/current/spec.html#Unions
    """
    for index, candidate in enumerate(schema):
        if validate(datum, candidate):
            break
    else:
        raise ValueError(
            '%r (type %s) does not match the schema: %s'
            % (datum, type(datum), schema)
        )
    write_long(stream, index)
    write_data(stream, datum, schema[index])


def write_record(stream, datum, schema):
    """A `record` value is written by encoding the values of its fields in the
    order in which they are declared. In other words, a `record` is written
    as just the concatenation of the encodings of its fields. Field values
    are encoded per their schema.

    Reference: https://avro.apache.org/docs/current/spec.html#schema_record
    """
    for field in schema['fields']:
        value = datum.get(field['name'], field.get('default', NoValue))
        write_data(stream, value, field['type'])


# ---- Validation of data with Schema ----------------------------------------#

def validate(datum, schema):
    """
    Determine if a python datum is an instance of a schema.

    This function is only called by `write_union`. Unfortunately, this function
    creates significant overhead for writing `union` types. Any ideas to
    improve this function would be especially welcomed.
    """

    if isinstance(schema, dict):
        record_type = schema['type']
    elif isinstance(schema, list):
        record_type = 'union'
    else:
        record_type = schema

    if record_type == 'null':
        return datum is None

    elif record_type == 'boolean':
        return isinstance(datum, bool)

    elif record_type == 'string':
        return isinstance(datum, _string_types)

    elif record_type == 'bytes':
        return isinstance(datum, _bytes_type)

    elif record_type == 'int':
        return (
            isinstance(datum, _int_types) and
            INT_MIN_VALUE <= datum <= INT_MAX_VALUE
        )

    elif record_type == 'long':
        return (
            isinstance(datum, _int_types) and
            LONG_MIN_VALUE <= datum <= LONG_MAX_VALUE
        )

    elif record_type in ('float', 'double',):
        return isinstance(datum, _number_types)

    elif record_type == 'fixed':
        return isinstance(datum, bytes) and len(datum) == schema['size']

    elif record_type in ('union', 'error_union',):
        return any(validate(datum, s) for s in schema)

    # dict-y types from here on.
    elif record_type == 'enum':
        return datum in schema['symbols']

    elif record_type == 'array':
        return (
            isinstance(datum, Iterable) and
            all(validate(d, schema['items']) for d in datum)
        )

    elif record_type == 'map':
        return (
            isinstance(datum, Mapping) and
            all(isinstance(k, _string_types) for k in datum.keys()) and
            all(validate(v, schema['values']) for v in datum.values())
        )

    elif record_type in ('record', 'error'):
        return (
            isinstance(datum, Mapping) and
            all(
                validate(datum.get(f['name'], f.get('default', NoValue)),
                         f['type'])
                for f in schema['fields']
            )
        )

    else:
        record_type = SCHEMA_DEFS.get(record_type)
        if record_type:
            return validate(datum, record_type)

    raise ValueError('Unknown record type: %s' % record_type)


# ---- Writer function lookup ------------------------------------------------#

WRITERS = {
    # Primitive types
    'null': write_null,
    'boolean': write_boolean,
    'int': write_int,
    'long': write_long,
    'float': write_float,
    'double': write_double,
    'bytes': write_bytes,
    'string': write_string,

    # Complex types
    'fixed': write_fixed,
    'enum': write_enum,
    'array': write_array,
    'map': write_map,
    'union': write_union,
    'record': write_record,
    'error': write_record,
    'error_union': write_union,
}


def write_data(stream, datum, schema):
    """Write a `datum` of data to the output `stream` using the specified
    Avro `schema`.

    Paramaters
    ----------
    stream: file-like object
        Output file or stream
    datum: object
        Data to write
    schema: dict
        Avro Schema for `datum`
    """
    if isinstance(schema, dict):
        record_type = schema['type']
    elif isinstance(schema, list):
        record_type = 'union'
    else:
        record_type = schema
    WRITERS[record_type](stream, datum, schema)


# ---- Block Encoders --------------------------------------------------------#

def null_write_block(stream, block_bytes):
    """Write a block of bytes with no codec ('null' codec)."""
    write_long(stream, len(block_bytes))
    stream.write(block_bytes)


def deflate_write_block(stream, block_bytes):
    """Write a block of bytes with the 'deflate' codec."""
    # The first two and last characters are zlib wrappers around deflate data
    data = compress(block_bytes)[2:-1]
    write_long(stream, len(data))
    stream.write(data)


def snappy_write_block(stream, block_bytes):
    """Write a block of bytes wih the 'snappy' codec."""
    data = snappy.compress(block_bytes)
    # Add 4 bytes for the CRC32
    write_long(stream, len(data) + 4)
    stream.write(data)
    # Write the 4-byte, big-endian CRC32 checksum
    crc = crc32(block_bytes) & 0xFFFFFFFF
    stream.write(pack('>I', crc))


# ---- Schema Handling -------------------------------------------------------#

SCHEMA_DEFS = dict((typ, typ) for typ in PRIMITIVE_TYPES)


def get_schema_defs():
    """Return the registered schema definitions."""
    return SCHEMA_DEFS


def acquaint_schema(schema, repo=None):
    """Extract `schema` into `repo` (default WRITERS)"""
    repo = WRITERS if repo is None else repo
    extract_named_schemas_into_repo(
        schema,
        repo,
        lambda schema: (
            lambda stream, datum, _: write_data(stream, datum, schema)
        ),
    )
    extract_named_schemas_into_repo(
        schema,
        SCHEMA_DEFS,
        lambda schema: schema,
    )


def write_header(stream, metadata, sync_marker):
    """Write the Avro header"""
    # Note: values in the `meta` dict are written as bytes.
    # See the definition of HEADER_SCHEMA in schema.py
    header = {
        'magic': MAGIC,
        'meta': dict(
            (k, v.encode('utf-8') if isinstance(v, _unicode_type) else v)
            for k, v in iteritems(metadata)
        ),
        'sync': sync_marker,
    }
    write_data(stream, header, HEADER_SCHEMA)


# ---- Public API - Writing Avro Files ---------------------------------------#

def writer(
    stream,
    schema,
    records,
    codec='null',
    sync_interval=SYNC_INTERVAL,
    metadata=None
):
    """Write multiple `records` to the output `stream` using the specified
    Avro `schema`.

    Paramaters
    ----------
    stream: file-like object
        Output file or stream
    schema: dict
        An Avro schema, typically a dict
    records: iterable
        Records to write
    codec: string, optional
        Compression codec, can be 'null', 'deflate' or 'snappy' (if installed)
    sync_interval: int, optional
        Size of sync interval. Defaults to fastavro.SYNC_INTERVAL
    metadata: dict, optional
        Additional header metadata

    Example
    -------
    >>> import fastavro
    >>>
    >>> schema = {
    >>>     'doc': 'A weather reading.',
    >>>     'name': 'Weather',
    >>>     'namespace': 'test',
    >>>     'type': 'record',
    >>>     'fields': [
    >>>         {'name': 'station', 'type': 'string'},
    >>>         {'name': 'time', 'type': 'long'},
    >>>         {'name': 'temp', 'type': 'int'},
    >>>     ],
    >>> }
    >>>
    >>> records = [
    >>>     {u'station': u'011990-99999', u'temp': 0, u'time': 1433269388},
    >>>     {u'station': u'011990-99999', u'temp': 22, u'time': 1433270389},
    >>>     {u'station': u'011990-99999', u'temp': -11, u'time': 1433273379},
    >>>     {u'station': u'012650-99999', u'temp': 111, u'time': 1433275478},
    >>> ]
    >>>
    >>> with open('weather.avro', 'wb') as out:
    >>>     fastavro.write(out, schema, records)
    """
    # Default values
    codec = codec or 'null'
    sync_interval = sync_interval or SYNC_INTERVAL
    metadata = metadata or {}

    # Get block writer specified by `codec`
    if codec == 'null':
        block_writer = null_write_block
    elif codec == 'deflate':
        block_writer = deflate_write_block
    elif codec == 'snappy':
        if not snappy:
            raise ValueError(
                "Cannot write 'snappy' codec: 'snappy' module is not available"
            )
        block_writer = snappy_write_block
    else:
        raise ValueError('Unknown codec: %r' % codec)

    # Write Avro header
    sync_marker = urandom(SYNC_SIZE)
    metadata['avro.codec'] = codec
    metadata['avro.schema'] = json.dumps(schema)
    write_header(stream, metadata, sync_marker)

    # Register the schema
    acquaint_schema(schema)

    buf = BytesIO()
    block_count = 0

    for record in records:
        write_data(buf, record, schema)
        block_count += 1
        if buf.tell() >= sync_interval:
            write_long(stream, block_count)
            block_writer(stream, buf.getvalue())
            stream.write(sync_marker)
            buf.truncate(0)
            buf.seek(0)
            block_count = 0

    if buf.tell() or block_count > 0:
        write_long(stream, block_count)
        block_writer(stream, buf.getvalue())
        stream.write(sync_marker)

    stream.flush()
    buf.close()


def schemaless_writer(stream, schema, record):
    """Write a single `record` to the output `stream` without writing the Avro
    schema and header information.

    Paramaters
    ----------
    stream: file-like object
        Output file or stream
    schema: dict
        Schema for `record`. (This is not written to the output stream,
        however it is required for writing the `record`.)
    record: dict
        Record to write
    """
    acquaint_schema(schema)
    write_data(stream, record, schema)
